### PR TITLE
[HIP] Correctly set HIP Kernel Buffer Arguments

### PR DESCRIPTION
**Pull request by veselypeta**
Device binary code exported from SYCL produces different interfaces depending on the backend selected. On CUDA and L0 buffer arguments are encoded with two parameters 1) A pointer to the buffer 2) a sycl::id struct (the accessor). However, on HIP the accessor argument is encoded as three individual i64 arguments.
This PR correctly sets the kernel arguments for kernels compiled and run with the HIP backend.

### CODE REVIEW

**Comment by ldrumm** [comment]
s/Compliation/Compilation/g

File: test/conformance/kernel/urKernelCreate.cpp [hunk 1]
```diff
@@ -28,12 +28,12 @@ struct urKernelCreateTest : uur::urProgramTest {
 UUR_INSTANTIATE_KERNEL_TEST_SUITE_P(urKernelCreateTest);
 
 TEST_P(urKernelCreateTest, Success) {
-    ASSERT_SUCCESS(urKernelCreate(program, kernel_name.data(), &kernel));
+    ASSERT_SUCCESS(urKernelCreate(program, kernel_name.c_str(), &kernel));
```
Comment (ldrumm) [inline test/conformance/kernel/urKernelCreate.cpp:31]: These changes seem unrelated

File: .github/workflows/cmake.yml [hunk 2]
```diff
@@ -180,7 +180,7 @@ jobs:
 
     - name: Download DPC++
       run: |
-        wget -O ${{github.workspace}}/dpcpp_compiler.tar.gz https://github.com/intel/llvm/releases/download/nightly-2023-09-21/sycl_linux.tar.gz
+        wget -O ${{github.workspace}}/dpcpp_compiler.tar.gz https://github.com/intel/llvm/releases/download/nightly-2023-10-18/sycl_linux.tar.gz
```
Comment (ldrumm) [inline .github/workflows/cmake.yml:183]: This change seems unrelated

File: test/conformance/device_code/CMakeLists.txt [hunk 3]
```diff
@@ -9,9 +9,13 @@ macro(add_device_binary SOURCE_FILE)
...
     foreach(TRIPLE ${TARGET_TRIPLES})
         set(EXE_PATH "${DEVICE_BINARY_DIR}/${KERNEL_NAME}_${TRIPLE}")
+        if(${TRIPLE} MATCHES "amd")
+            set(AMD_TARGET_BACKEND -Xsycl-target-backend=${TRIPLE})
+            set(AMD_OFFLOAD_ARCH  --offload-arch=${AMD_ARCH})
+        endif()
         add_custom_command(OUTPUT ${EXE_PATH}
-            COMMAND ${UR_DPCXX} -fsycl -fsycl-targets=${TRIPLE} -fsycl-device-code-split=off
-            ${SOURCE_FILE} -o ${EXE_PATH}
+            COMMAND ${UR_DPCXX} -fsycl -fsycl-targets=${TRIPLE} -fsycl-device-code-split=off ${AMD_TARGET_BACKEND} ${AMD_OFFLOAD_ARCH}
```
Comment (ldrumm) [inline test/conformance/device_code/CMakeLists.txt:17]: this line's getting a bit long

File: source/adapters/hip/device.cpp [hunk 4]
```diff
@@ -30,6 +30,9 @@ UR_APIEXPORT ur_result_t UR_APICALL urDeviceGetInfo(ur_device_handle_t hDevice,
   static constexpr uint32_t MaxWorkItemDimensions = 3u;
 
   switch ((uint32_t)propName) {
+  case UR_DEVICE_INFO_VIRTUAL_MEMORY_SUPPORT: {
```
Comment (ldrumm) [inline source/adapters/hip/device.cpp:33]: Is this related?

**Review by ldrumm — CHANGES_REQUESTED** [review]
Some changes don't look like they belong here. Please clarify their purpose, or remove unrelated cleanups

File: test/conformance/kernel/urKernelCreate.cpp (hunk 1 above)
Comment (veselypeta) [reply test/conformance/kernel/urKernelCreate.cpp]: 👍

File: test/conformance/device_code/CMakeLists.txt (hunk 3 above)
Comment (veselypeta) [reply test/conformance/device_code/CMakeLists.txt]: 👍

File: source/adapters/hip/device.cpp (hunk 4 above)
Comment (veselypeta) [reply source/adapters/hip/device.cpp]: Not directly 👍 - will include in a separate patch

File: test/conformance/source/environment.cpp [hunk 5]
```diff
@@ -266,6 +266,17 @@ std::string KernelsEnvironment::getSupportedILPostfix(uint32_t device_index) {
         return {};
     }
 
+    // special case for AMD as it doesn't support IL.
```
Comment (pbalcer) [inline test/conformance/source/environment.cpp:269]: Is this a TODO? should be there an issue to fix IL then?

File: test/conformance/device_code/CMakeLists.txt [hunk 6]
```diff
@@ -9,9 +9,15 @@ macro(add_device_binary SOURCE_FILE)
...
+        if(${TRIPLE} MATCHES "amd")
+            set(AMD_TARGET_BACKEND -Xsycl-target-backend=${TRIPLE})
+            set(AMD_OFFLOAD_ARCH  --offload-arch=${AMD_ARCH})
+        endif()
         add_custom_command(OUTPUT ${EXE_PATH}
-            COMMAND ${UR_DPCXX} -fsycl -fsycl-targets=${TRIPLE} -fsycl-device-code-split=off
-            ${SOURCE_FILE} -o ${EXE_PATH}
+            COMMAND ${UR_DPCXX} -fsycl -fsycl-targets=${TRIPLE} -fsycl-device-code-split=off 
+            ${AMD_TARGET_BACKEND} ${AMD_OFFLOAD_ARCH} ${SOURCE_FILE} 
+            -o ${EXE_PATH} || (exit 0)
```
Comment (pbalcer) [inline test/conformance/device_code/CMakeLists.txt:19]: Ignoring the error on dumping the device binaries already makes it easy to accidentally misconfigure these tests. Ignoring compiler errors will only make the problem worse. Can you please add a comment why the errors are ignored?

I'm wondering whether it would be possible to check if running `${EXE_PATH}` created a file in cwd. Maybe with inotify?

**Comment by pbalcer** [comment]
Closes #842

File: test/conformance/source/environment.cpp (hunk 5 above)
Comment (veselypeta) [reply test/conformance/source/environment.cpp:269]: Possibly, although from speaking to the team implementing the plugin they are likely to not support this query as the HIP does not accept 'IL' as such.

File: test/conformance/device_code/CMakeLists.txt (hunk 6 above)
Comment (veselypeta) [reply test/conformance/device_code/CMakeLists.txt:19]: This is a slightly strange issue I've found. AMD does not support images (at least through SYCL) and so the compilation of `image_copy.cpp` fails. I was using Ninja as a generator locally and it's perfectly happy to ignore this error and let the build continue, but using "Unix Makefiles" stops the build and exits. Perhaps we should just skip compiling this code on AMD and avoid having to silence the error. I've already added checks to the image/sampler CTS to ensure that this program is not loaded on AMD. 👍

**Review by ldrumm — APPROVED** [review]

**Review by pbalcer — APPROVED** [review]

File: test/conformance/enqueue/enqueue_adapter_hip.match [hunk 7]
```diff
@@ -1 +1,87 @@
-Segmentation fault
+{{OPT}}Segmentation Fault
```
[the remark this answers is not in the document]
Comment (veselypeta) [reply test/conformance/enqueue/enqueue_adapter_hip.match:1]: It seems to be related to what hardware the conformance tests are running on. I've created #985 to track this. I still think this is an improvement as none of these tests were being run on AMD before.

File: test/conformance/device_code/CMakeLists.txt [hunk 8]
```diff
@@ -9,9 +9,18 @@ macro(add_device_binary SOURCE_FILE)
     file(MAKE_DIRECTORY ${DEVICE_BINARY_DIR})
     foreach(TRIPLE ${TARGET_TRIPLES})
         set(EXE_PATH "${DEVICE_BINARY_DIR}/${KERNEL_NAME}_${TRIPLE}")
+        if(${TRIPLE} MATCHES "amd")
+            set(AMD_TARGET_BACKEND -Xsycl-target-backend=${TRIPLE})
+            set(AMD_OFFLOAD_ARCH  --offload-arch=${AMD_ARCH})
+        endif()
+        if(${TRIPLE} MATCHES "amd" AND ${KERNEL_NAME} MATCHES "image_copy")
+            continue()
```
[the remark this answers is not in the document]
Comment (veselypeta) [reply test/conformance/device_code/CMakeLists.txt:18]: I agree. Not sure what the best solution is overall. In intel/llvm the e2e tests that require sampler and image support specifically exclude amd with `UNSUPPORTED:` https://github.com/intel/llvm/blob/sycl/sycl/test-e2e/Sampler/unnormalized-none-nearest.cpp#L2

**Review by pbalcer — APPROVED** [review]

**Comment by jinz2014** [comment]
@veselypeta  Do you know the cause of the differences in accessor arguments between HIP and CUDA ?  Thanks.

**Comment by veselypeta** [comment]
Covered in #1030
